src/sources/http_scrape/scrape.rs

```diff
@@ -0,0 +1,243 @@
+            scrape_interval_secs,
+            query,
+            decoding,
+            framing,
+            headers,
+            tls,
+            auth,
+        }
+    }
+}
+
+inventory::submit! {
+    SourceDescription::new::<HttpScrapeConfig>(NAME)
+}
+
+impl_generate_config_from_default!(HttpScrapeConfig);
+
+#[async_trait::async_trait]
+#[typetag::serde(name = "http_scrape")]
+impl SourceConfig for HttpScrapeConfig {
+    async fn build(&self, cx: SourceContext) -> Result<sources::Source> {
+        // build the url
+        let endpoints = vec![self.endpoint.clone()];
+        let urls = endpoints
+            .iter()
+            .map(|s| s.parse::<Uri>().context(sources::UriParseSnafu))
+            .map(|r| r.map(|uri| super::get_url(&uri, &self.query)))
+            .collect::<std::result::Result<Vec<Uri>, sources::BuildError>>()?;
+
+        let tls = TlsSettings::from_options(&self.tls)?;
+
+        // build the decoder
+        let decoder = DecodingConfig::new(
+            self.framing.clone(),
+            self.decoding.clone(),
+            LogNamespace::Vector,
+        )
+        .build();
+
+        let content_type = self.decoding.content_type(&self.framing).to_string();
+
+        // the only specific context needed is the codec decoding
+        let context = HttpScrapeContext { decoder };
+
+        let inputs = super::GenericHttpScrapeInputs {
+            urls,
+            interval_secs: self.scrape_interval_secs,
+            headers: self.headers.clone(),
+            content_type,
+            auth: self.auth.clone(),
+            tls,
+            proxy: cx.proxy.clone(),
+            shutdown: cx.shutdown,
+        };
+
+        Ok(super::http_scrape(inputs, context, cx.out).boxed())
+    }
+
+    fn outputs(&self, _global_log_namespace: LogNamespace) -> Vec<Output> {
+        vec![Output::default(self.decoding.output_type())]
+    }
+
+    fn source_type(&self) -> &'static str {
+        NAME
+    }
+
+    fn can_acknowledge(&self) -> bool {
+        false
+    }
+}
+
+#[derive(Clone)]
+struct HttpScrapeContext {
+    decoder: Decoder,
+}
+
+impl HttpScrapeContext {
+    /// Decode the events from the byte buffer
+    fn decode_events(&mut self, buf: &mut BytesMut) -> Vec<Event> {
+        let mut events = Vec::new();
+        loop {
+            match self.decoder.decode_eof(buf) {
+                Ok(Some((next, _))) => {
+                    events.extend(next.into_iter());
+                }
+                Ok(None) => break,
+                Err(error) => {
+                    // Error is logged by `crate::codecs::Decoder`, no further
+                    // handling is needed here.
+                    if !error.can_continue() {
+                        break;
+                    }
+                    break;
+                }
+            }
+        }
+        events
+    }
+
+    /// Enriches events with source_type, timestamp
+    fn enrich_events(&self, events: &mut Vec<Event>) {
```

Comment:
Not a big deal, but this doesn't seem to need `&self`. It could just be a standalone function.